source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,44 @@
 class UrlsController < ApplicationController
+  def index
+    # for now, lets grab them all and pass them to a display page
+    @urls = Url.all
+  end
+
+  def new
+    # default behavior looks good
+    @target = ''
+    @save_errors = nil
+  end
+
+  def _extract_url_errors(url)
+    index = [:target_link, :linkid, :base].find { |i| url.errors[i].any? }
+    return nil if index.nil?
+    url.errors[index]
+  end
+
+  def create
+    # Actually create the new shortened url
+    @target = params[:target_link]
+    @errors = nil
+
+    @url = Url.new(target_link: @target)
+    if !@url.save
```

Comment:
I prefer the happy path as the positive part of the `if` and the error path under `else`.